caput/pipeline.py

```diff
@@ -593,11 +614,17 @@ def run(self):
         This function initializes all pipeline tasks and runs the pipeline
         through to completion.
 
+        Raises
+        ------
+        PipelineRuntimeError
+            If a task stage returns the wrong number of outputs.
         """
-
-        # Run the pipeline.
         while self.tasks:
             for task in list(self.tasks):  # Copy list so we can alter it.
+                if self._psutil_profiling:
```

Comment:
i.e. just create a new instance in here every time.

Reply:
I don't think running psutil has much overhead at all. I would be very tempted to run it *every* time (needing `psutil` to be an actual dependency), and have the only thing that changes be whether it actually outputs the stats.